axelrod/tests/unit/test_tournament.py

```diff
@@ -394,30 +397,128 @@ def test_play_matches(self, turns):
             players=self.players,
             game=self.game,
             repetitions=self.test_repetitions)
-        matches = {
-            (0, 0): axelrod.Match((axelrod.Cooperator(), axelrod.Cooperator()), turns=turns),
-            (0, 1): axelrod.Match((axelrod.Cooperator(), axelrod.TitForTat()), turns=turns),
-            (0, 2): axelrod.Match((axelrod.Cooperator(), axelrod.Defector()), turns=turns),
-            (0, 3): axelrod.Match((axelrod.Cooperator(), axelrod.Grudger()), turns=turns),
-            (0, 4): axelrod.Match((axelrod.Cooperator(), axelrod.GoByMajority()), turns=turns),
-            (1, 1): axelrod.Match((axelrod.TitForTat(), axelrod.TitForTat()), turns=turns),
-            (1, 2): axelrod.Match((axelrod.TitForTat(), axelrod.Defector()), turns=turns),
-            (1, 3): axelrod.Match((axelrod.TitForTat(), axelrod.Grudger()), turns=turns),
-            (1, 4): axelrod.Match((axelrod.TitForTat(), axelrod.GoByMajority()), turns=turns),
-            (2, 2): axelrod.Match((axelrod.Defector(), axelrod.Defector()), turns=turns),
-            (2, 3): axelrod.Match((axelrod.Defector(), axelrod.Grudger()), turns=turns),
-            (2, 4): axelrod.Match((axelrod.Defector(), axelrod.GoByMajority()), turns=turns),
-            (3, 3): axelrod.Match((axelrod.Grudger(), axelrod.Grudger()), turns=turns),
-            (3, 4): axelrod.Match((axelrod.Grudger(), axelrod.GoByMajority()), turns=turns),
-            (4, 4): axelrod.Match((axelrod.GoByMajority(), axelrod.GoByMajority()), turns=turns),
-        }
-        tournament._play_matches(matches)
-        for m in matches.values():
-            self.assertEqual(len(m), turns)
-            # Check that the name of the winner is in the list of names of
-            # players in the tournament (the way the matches are created these
-            # are not the same players).
-            self.assertIn(str(m.winner()), [str(p) for p in tournament.players] + ['False'])
+
+        player_classes = [axelrod.Cooperator, axelrod.TitForTat,
+                          axelrod.Defector, axelrod.Grudger]
+        matches = []
+        for i, player_cls in enumerate(player_classes):
+            for j, opponent_cls in enumerate(player_classes):
+                players = (player_cls(), opponent_cls())
+                match = axelrod.Match(players, turns=turns)
+                matches.append(((i, j), match))
+        matches_generator = (element for element in matches)
+
+        self.assertEqual((len(list(matches_generator))), len(matches))
+
+        interactions = tournament._play_matches(matches_generator)
```

Comment:
With a generator, once you've exhausted it once it usually raises `StopIterationError` and should give an empty list. However the generator function could be programmed to reset, like `range`:
```
>>> gen = range(0, 5)
>>> print(list(gen))
[0, 1, 2, 3, 4]
>>> print(list(gen))
[0, 1, 2, 3, 4]
```
Contrast with:
```
>>> gen = (i for i in range(0, 5))
>>> print(list(gen))
[0, 1, 2, 3, 4]
>>> print(list(gen))
[]
```
`_play_matches` should work with either a generator or a list, right? If the point is to test with a generator, let's simply put a comment above the generator creation block so that it's not confusing in the future.